### PR TITLE
Add RankSEG integration tutorial

### PR DESCRIPTION
**Pull request by statmlben**
Related to [MONAI-#8908](https://github.com/Project-MONAI/MONAI/pull/8908#issuecomment-5077397523)

### Description

This PR adds a runnable tutorial demonstrating how to use RankSEG as an optional third-party post-processing transform in a MONAI workflow.

The tutorial:

- uses the pretrained `pancreas_ct_dints_segmentation` MONAI Bundle and a real MSD Task07 Pancreas case;
- compares conventional argmax decoding with RankSEG using per-class Dice scores and a visualization;
- demonstrates the array-based `RankSEG` transform;
- demonstrates the dictionary-based `RankSEGd` transform in a MONAI `Compose` post-processing pipeline;
- documents the public dataset source, CC BY-SA 4.0 license, and research-use context.

This PR also adds the tutorial to the repository README and registers it in `doesnt_contain_max_epochs` because it is an inference-only notebook.

### Checks

- [x] Avoid including large-size files in the PR.
- [x] Clean up long text outputs from code cells in the notebook.
- [x] Check the contents and remove sensitive information such as user names and private keys.
- [x] Verify hyperlinks, markdown content, and relative repository paths.
- [x] Notebook runs automatically with:

```bash
./runner.sh -t modules/rankseg_integration.ipynb
```

### Results

```
Argmax macro Dice: 0.7747
RankSEG macro Dice: 0.8502
Paired Dice change: +0.0755
Pancreas Dice: 0.8764 -> 0.8898
Tumor Dice: 0.6730 -> 0.8106
```

image

## Summary by CodeRabbit

* **New Features**
  * Added a tutorial for optional RankSEG/RankSEGd post-processing in MONAI segmentation workflows.
  * Compares RankSEG results with standard argmax predictions using a pretrained pancreas segmentation model.
  * Includes Dice score evaluation, visual comparisons, and tensor- and dictionary-based workflow examples.
  * Demonstrates dataset preparation and integration validation.

* **Documentation**
  * Added the RankSEG integration notebook to the list of available modules and examples.

### CODE REVIEW

File: modules/rankseg_integration.ipynb [hunk 1]
```diff
@@ -0,0 +1,2104 @@
...
+    "    dints_space=topology,\n",
+    "    in_channels=1,\n",
+    "    num_classes=3,\n",
+    "    use_downsample=True,\n",
+    "    node_a=torch.as_tensor(architecture[\"node_a\"], device=device),\n",
+    ").to(device)\n",
+    "checkpoint = torch.load(\n",
+    "    os.path.join(bundle_dir, \"models\", \"model.pt\"),\n",
+    "    map_location=device,\n",
+    "    weights_only=False,\n",
```
[the remark this answers is not in the document]
Comment (ericspod) [reply modules/rankseg_integration.ipynb:675]: This is an issue with the weights being stored using Numpy arrays rather than tensors. I have raised this here: https://github.com/Project-MONAI/MONAI/issues/9025 For now we'll stick with the current loading strategy as we know these files are safe.

File: modules/rankseg_integration.ipynb [hunk 2]
```diff
@@ -0,0 +1,2104 @@
...
+    "    arch_code=[architecture[\"arch_code_a\"], architecture[\"arch_code_c\"]],\n",
+    "    device=str(device),\n",
+    ")\n",
+    "model = DiNTS(\n",
+    "    dints_space=topology,\n",
+    "    in_channels=1,\n",
+    "    num_classes=3,\n",
+    "    use_downsample=True,\n",
+    "    node_a=torch.as_tensor(architecture[\"node_a\"], device=device),\n",
+    ").to(device)\n",
```
Comment (ericspod) [inline modules/rankseg_integration.ipynb]: It's possible to use the bundle parser to load the model for you:

```python
p = monai.bundle.ConfigParser()
p.read_config(bundle_dir + "/configs/inference.yaml")
p["arch_ckpt"] = architecture
model = p.get_parsed_content("network_def")
```

File: modules/rankseg_integration.ipynb [hunk 3]
```diff
@@ -0,0 +1,2104 @@
...
+    "execution": {
+     "iopub.execute_input": "2026-07-27T08:23:21.500843Z",
+     "iopub.status.busy": "2026-07-27T08:23:21.500368Z",
+     "iopub.status.idle": "2026-07-27T08:23:24.891073Z",
+     "shell.execute_reply": "2026-07-27T08:23:24.890402Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "!python -c \"import monai\" || pip install -q \"monai-weekly[nibabel, tqdm]\"\n",
```
[the remark this answers is not in the document]
Comment (ericspod) [reply modules/rankseg_integration.ipynb]: An alternative is to use `%pip install -q --upgrade-strategy only-if-needed "monai-weekly[nibabel, tqdm,matplotlib,huggingface_hub]" rankseg` in place of the multiple `pip` commands. This won't install over top of existing MONAI installs but will install missing dependencies.

**Review by ericspod — CHANGES_REQUESTED** [review]
Hi @statmlben thanks for this tutorial as we had discussed. I think it's fine overall for the existing content, but we need a discussion on what Rankseg is and how the algorithm works. At the top you should describe what it is, how it differs from other components, what it's used for, what the implications of using it are, etc. This by no means needs the same details as in your paper but should be motivating enough so readers know what the notebook is demonstrating and whether they'd want to use it themselves or not. I think the Coderabbit comments should be looked at but I had responses here as well.

File: modules/rankseg_integration.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (statmlben) [reply modules/rankseg_integration.ipynb]: Thanks, agree. I replaced the separate conditional installation commands with the single kernel-aware `%pip` command you suggested. The full notebook runner passes with this change.

File: modules/rankseg_integration.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (statmlben) [reply modules/rankseg_integration.ipynb:675]: Thanks for clarifying and for opening Project-MONAI/MONAI#9025. I kept the current `weights_only=False` loading strategy for these official Bundle artifacts as suggested.

File: modules/rankseg_integration.ipynb (hunk 2 above)
Comment (statmlben) [reply modules/rankseg_integration.ipynb]: Thanks for the suggestion. I updated the notebook to construct `network_def` from the Bundle configuration using `ConfigParser`.

I also override `dints_space#device` with the selected runtime device because the current Bundle configuration hardcodes CUDA, while the tutorial retains a CPU fallback. The full notebook runner passes with this change.

**Comment by statmlben** [comment]
Thanks @ericspod for the feedback. I added an introductory section near the top of the notebook explaining what RankSEG is, how it differs from argmax, fixed thresholding, and morphology-based post-processing, and how the RMA workflow operates at a high level. The section also discusses suitable use cases, probability requirements, computational cost, and the need to validate improvements on representative data. 

I also addressed your inline suggestions by consolidating the dependency installation command and using the Bundle `ConfigParser` to construct the model.

File: modules/rankseg_integration.ipynb [hunk 4]
```diff
@@ -0,0 +1,2121 @@
...
+     ]
+    }
+   ],
+   "source": [
+    "case_id = \"pancreas_102\"\n",
+    "data_dir = os.path.join(root_dir, \"Task07_Pancreas\")\n",
+    "evaluation_file = {}\n",
+    "for key, subdirectory in ((\"image\", \"imagesTr\"), (\"label\", \"labelsTr\")):\n",
+    "    evaluation_file[key] = hf_hub_download(\n",
+    "        repo_id=\"MedOtter/msd-pancreas\",\n",
```
Comment (garciadias) [inline modules/rankseg_integration.ipynb]: This downloads the MSD Task07 Pancreas case from repo_id="MedOtter/msd-pancreas", a personal/community Hugging Face dataset repo with no MONAI or MSD organizational affiliation and no checksum verification. Every other MSD download in this repo goes through a MONAI-controlled source instead: the msd-for-monai S3 bucket (e.g. spleen_segmentation_3d.ipynb), monai.apps.DecathlonDataset, or an official MONAI-org HF repo (hugging_face/finetune_vista3d_for_hugging_face_pipeline.ipynb uses MONAI/VISTA3D-HF). Since this tutorial is meant to stay in the repo long-term, relying on one individual's unmanaged mirror is a real breakage/provenance risk if that repo is edited or removed.

I checked, and there isn't yet a first-party equivalent that offers a single downloadable Task07 Pancreas case: neither the MONAI nor Project-MONAI Hugging Face orgs host any MSD Task07 dataset, and monai.apps.DecathlonDataset only supports downloading and extracting the full archive (no single-case mode). That said, MONAI's own official mirror at https://msd-for-monai.s3-us-west-2.amazonaws.com/Task07_Pancreas.tar (the same URL DecathlonDataset uses internally) does support HTTP range requests — I confirmed Accept-Ranges: bytes and a 206 Partial Content response to a ranged GET against it — so it is technically possible to fetch just the pancreas_102 case from this first-party source via a targeted tar-header walk, without downloading the full archive. Note the full tar is about 11.4 GB (Content-Length: 12289971712), not ~2 GB as the PR description implies, so a full-archive download via DecathlonDataset would be a much larger tradeoff than expected.

Suggested change: Prefer MONAI's own S3 mirror (msd-for-monai.s3-us-west-2.amazonaws.com/Task07_Pancreas.tar) over the personal HF mirror — either via a small range-request helper that reads the tar's header table to extract only pancreas_102, or by accepting the larger DecathlonDataset download if that's simpler to maintain. If neither is workable for this PR, at minimum please document the provenance/maintenance risk of the current mirror in the notebook and consider it a stopgap rather than a long-term solution.

File: modules/rankseg_integration.ipynb [hunk 5]
```diff
@@ -0,0 +1,2121 @@
...
+    "execution": {
+     "iopub.execute_input": "2026-07-27T08:23:21.500843Z",
+     "iopub.status.busy": "2026-07-27T08:23:21.500368Z",
+     "iopub.status.idle": "2026-07-27T08:23:24.891073Z",
+     "shell.execute_reply": "2026-07-27T08:23:24.890402Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "%pip install -q --upgrade-strategy only-if-needed \"monai-weekly[nibabel, tqdm,matplotlib,huggingface_hub]\" rankseg\n",
```
Comment (garciadias) [inline modules/rankseg_integration.ipynb]: rankseg is installed unpinned here, but it's the tutorial's central third-party dependency and is currently pre-1.0 (0.0.5 on PyPI), so its API (e.g. the pruning_prob and output_mode kwargs this notebook relies on) is more likely to change than a stable library. Other notebooks in this repo pin non-MONAI dependencies for this reason (e.g. torchio==0.18.73, pytorch-lightning==1.5.10, scikit-learn==1.0.2).

Suggested change: Pin to the version this notebook was verified against, e.g. rankseg==0.0.5.

File: modules/rankseg_integration.ipynb [hunk 6]
```diff
@@ -0,0 +1,2121 @@
...
+    "- In the fixed 20-case comparison from which this case was selected, RankSEG (Dice) changed mean samplewise Dice by `+0.0209` and won on 18 of 20 cases. Use that paired evaluation, rather than this selected example alone, for performance claims.\n",
+    "- This wrapper intentionally demonstrates non-overlapping multiclass output. Multilabel RankSEG returns `(C, *spatial)` binary masks after removing the temporary batch dimension and needs slightly different shape handling.\n",
+    "- The Bundle and this tutorial are research examples and are not intended for diagnostic use.\n",
+    "\n",
+    "Further reading:\n",
+    "\n",
+    "- [RankSEG repository and documentation](https://github.com/rankseg/rankseg)\n",
+    "- [RankSEG: A Consistent Ranking-based Framework for Segmentation](https://www.jmlr.org/papers/v24/22-0712.html)\n",
+    "- [RankSEG-RMA: An Efficient Segmentation Algorithm via Reciprocal Moment Approximation](https://openreview.net/forum?id=4tRMm1JJhw)\n",
+    "- [Reproducible MONAI Pancreas comparison](https://github.com/rankseg/rankseg/tree/main/experiments/monai-pr-8908-reproducibility)\n",
```
Comment (garciadias) [inline modules/rankseg_integration.ipynb]: The "Practical notes" cell states a specific external benchmark (+0.0209 mean samplewise Dice, 18/20 cases) and links to rankseg/rankseg on the main branch. Since main can move, this claim isn't reproducibly tied to a fixed artifact.

Suggested change: Link to a tagged release or fixed commit SHA of that experiments directory instead of main.

**Review by garciadias — CHANGES_REQUESTED** [review]
Thanks for this tutorial — I built a GPU Docker container (monai:blackwell, RTX 5090), installed the notebook's declared dependencies exactly as specified, and ran the full `./runner.sh -t modules/rankseg_integration.ipynb` end-to-end. It passed cleanly and reproduced the PR description's numbers exactly (Argmax macro Dice 0.7747, RankSEG macro Dice 0.8502, Pancreas 0.8764 -> 0.8898, Tumor 0.6730 -> 0.8106). The RankSEG/RankSEGd transform code, shape handling, and Dice evaluation are all correct.

I left three inline comments. The main one (major) is that the MSD Task07 Pancreas case is downloaded from an unofficial personal Hugging Face account (`MedOtter/msd-pancreas`) rather than a MONAI-controlled source, unlike every other MSD download in this repo — I've included a concrete alternative (MONAI's own S3 mirror, which supports HTTP range requests) in that comment. The other two are minor (pinning the rankseg dependency, and linking to a fixed commit rather than a moving branch for an external benchmark claim).

File: modules/rankseg_integration.ipynb (hunk 4 above)
Comment (statmlben) [reply modules/rankseg_integration.ipynb]: Thanks for pointing out the provenance and maintenance risk. I replaced the personal Hugging Face mirror with HTTP Range requests against a fixed object version of MONAI's official msd-for-monai S3 archive.

The helper downloads only the tar headers and contents for pancreas_102 and its label (about 29.7 MB total). It verifies the object ETag, tar member `checksum/name/size/type`, and SHA-256 digests. I also reran the complete notebook successfully with this source.

File: modules/rankseg_integration.ipynb (hunk 5 above)
Comment (statmlben) [reply modules/rankseg_integration.ipynb]: Agreed. I pinned the dependency to rankseg==0.0.5, which is the exact version used to execute and verify the notebook.

File: modules/rankseg_integration.ipynb (hunk 6 above)
Comment (statmlben) [reply modules/rankseg_integration.ipynb]: Agreed. I replaced the moving main-branch link with a link to the fixed commit that introduced the reproducibility experiment:

[https://github.com/rankseg/rankseg/tree/2f9e7e9e36145d960d3b568f340c4ccb60b397d4/experiments/monai-pr-8908-reproducibility](https://github.com/rankseg/rankseg/tree/2f9e7e9e36145d960d3b568f340c4ccb60b397d4/experiments/monai-pr-8908-reproducibility)

**Comment by statmlben** [comment]
Hi @garciadias , thank you for the detailed review. I addressed all three comments by switching to MONAI's official MSD S3 source with integrity checks, pinning `rankseg==0.0.5`, and linking the benchmark to a fixed commit.

Thanks again for helping improve the tutorial's provenance and reproducibility

**Review by garciadias — APPROVED** [review]
All three review findings are resolved:

- The MSD Task07 Pancreas case now downloads from MONAI's own official S3 mirror (msd-for-monai.s3-us-west-2.amazonaws.com), pinned to an immutable S3 object version, via HTTP range requests that validate the tar header (name, size, type) and verify a SHA-256 digest of the extracted bytes — a more rigorous fix than what I suggested.
- rankseg is now pinned to ==0.0.5.
- The reproducibility link now points to a fixed commit SHA instead of the main branch.

I re-verified by rebuilding a clean GPU Docker container (monai:blackwell, RTX 5090) against a fresh worktree at the latest commit, installing the notebook's declared dependencies exactly as specified, and running the full `./runner.sh -t modules/rankseg_integration.ipynb` end-to-end. It passed and reproduced identical results (Argmax macro Dice 0.7747, RankSEG macro Dice 0.8502, Pancreas 0.8764 -> 0.8898, Tumor 0.6730 -> 0.8106). I also independently confirmed the downloaded image and label files' SHA-256 hashes match the values hardcoded in the notebook. LGTM.

**Comment by statmlben** [comment]
Thanks @garciadias for the thorough re-review and independent verification, much appreciated!

Hi @ericspod, I’ve addressed the requested changes, including the introductory discussion of RankSEG and the dependency/model-loading updates. @garciadias has also re-run the complete notebook in a clean GPU environment and approved the PR. The remaining Actions checks were cancelled by GitHub infrastructure errors before the repository code was executed (`hosted runner not acquired` and `Service Unavailable` while resolving action downloads). Could you please re-review the changes and, if possible, re-run the affected workflows? Thanks!

**Comment by garciadias** [comment]
> Thanks @garciadias for the thorough re-review and independent verification, much appreciated!
> 
> Hi @ericspod, I’ve addressed the requested changes, including the introductory discussion of RankSEG and the dependency/model-loading updates. @garciadias has also re-run the complete notebook in a clean GPU environment and approved the PR. The remaining Actions checks were cancelled by GitHub infrastructure errors before the repository code was executed (`hosted runner not acquired` and `Service Unavailable` while resolving action downloads). Could you please re-review the changes and, if possible, re-run the affected workflows? Thanks!

Sorry @statmlben, somehow GitHub failed to run some of the CI, and this got stuck for a while. Now that everything is passing, we can just wait for @ericspod to approve and merge.

**Review by ericspod — APPROVED** [review]
